### ISSUE TITLE
message.publish hangs forever sometimes

### ISSUE THREAD

**Issue by gilbsgilbs** [mosquito/aio-pika#46]
Hi!

It happens that `message.publish` hangs forever although the message was successfully published. See [here](https://github.com/mosquito/aio-pika/issues/46#issuecomment-316130980).

~~It happens that I `reject` messages at the end of a `basic_consume` callback. Though, when I call  `reject()`, it happens that messages are not actually rejected, and just remain in unacked state until I close the connection.~~

~~As soon as I add logging around `reject()`, it starts working again. Do you see any reason why this is happening and how to workaround this ?~~

~~I've been digging into `pika` and in `BaseConnection`, there's the `_handle_write` function that is supposed to write to the socket. As you explained in #7 this might be non-blocking (although I don't see why, because it seems to write to the socket explicitly and expects timeout ; there must be a gotcha). Any chance that this is not flushed before the end of the task under some circumstances?~~

Thanks!
Gilb's

**Comment by mosquito** [comment]
Try to capture traffic with (`wireshark` or `tcpdump -pni eth0 -w /tmp/traff.pcap`). For me it's so useful strategy. Wireshark able to decode amqp protocol.

**Comment by gilbsgilbs** [comment]
I think that's a brilliant idea. Thanks.

**Comment by gilbsgilbs** [comment]
After digging a bit more, it appears that the problem is a bit different than what I thought in the first place.

`reject` is indeed not called, but that's because I `await message.publish(…)` just before, and **this** seems to hang forever for no reason.

The last (and only) log produced by `aio_pika` is:
```
Publishing message via exchange test.xch: Message:{'app_id': None,
'body_size': 1222,
'content_encoding': None,
'content_type': None,
'correlation_id': b'None',
'delivery_mode': 2,
'expiration': 120,
'headers': {'x-death': [{'count': 2L,
'exchange': 'test.xch',
'expiration': '30000',
'queue': 'test.queue',
'reason': 'expired',
'routing-keys': [''],
'time': datetime.datetime(2017, 7, 18, 16, 36, 50)}]},
'message_id': None,
'priority': None,
'reply_to': None,
'timestamp': None,
'type': 'None',
'user_id': None}
```

It's as if `aio_pika` never received the `confirm_delivery` callback and just waits for it forever. Although, I can see through wireshark that the message was sent.

Just to mention, I have a ELB in front of RabbitMQ. Don't think it has any correlation, but I might be wrong.

**Comment by mosquito** [comment]
Can you temporary disable delivery confirmation for your `aio-pika` installation? Just for test of course.

**Comment by gilbsgilbs** [comment]
@mosquito It seems to behave normally when I disable delivery confirmation.
(commented `channel.confirm_delivery(…)` and set the future result manually in `_publish(…)` in `channel.py`)

Though, before disabling `confirm_delivery`, I would expect `Basic.Ack` to appear in wireshark when `publishing`, which is NOT the case when the issue occurs.

I'm still trying hard to reproduce the issue with a tiny code snippet.

**Comment by mosquito** [comment]
Hi @gilbsgilbs in #50 @smagafurov adding flag for disabling of publisher confirmations. It's like workaround out for this issue. But I still waiting for an example which will repruduce it. I hope you can do this. stabilization of the `aio-pika` codebase it's my aim.

**Comment by gilbsgilbs** [comment]
Hi!
After spending a bit more time on this today, I finally figured out what was happening. Good news: it was not aio-pika's fault at all but mine.

I was trying to reproduce the bug on the faulty remote server, and noticed I had several similar issues that I wasn't able to reproduce locally (e.g. `aio-pika` hanging one time out of two (literally) while declaring queues). Oddly enough, I never managed to reproduce the issue described in the first post. I tried to update the remote python to 3.6 to ensure it was not `asyncio` related, but appearently it was not. Then, I just thought what could be the differences between my local machine vs remote. And I checked AMQP version, and wow, the remote version was almost 1.5 years old. Yeah, our PPA GPG key was outdated (or has always been wrong), which made Ubuntu ignore `rabbitmq-server` updates for that long (without any complaint in our orchestration software) 😖 . Apparently, the bug was fixed in a later release of RabbitMQ.

Two lessons learned:
- Always check for outdated version of software even when you're 100% sure it's up-to-date. It might not be.
- Ensure that your orchestration tool does not mute stderr during sysupgrades 😠 

Sorry for the time loss. I should have noticed earlier that remote RabbitMQ was outdated.

Gilb's
